File: ts-src/commands/builtins.ts
import { loadHistory } from "../history/reader.ts";
import { listPlans, loadPlan } from "../plans/storage.ts";
import { formatTasksSummary, loadTasks, saveTasks } from "../tasks/storage.ts";
import type { CommandRegistry } from "./registry.ts";
import type { SlashCommand } from "./types.ts";

export function createBuiltinCommands(commandRegistry: CommandRegistry): SlashCommand[] {
	return [
		{
			name: "help",
			description: "List available commands",
			execute: async () => {
				for (const cmd of commandRegistry.all()) {
					console.log(`  /${cmd.name} — ${cmd.description}`);
				}
			},
		},
		{
			name: "clear",
			description: "Clear conversation context",
			execute: async (_args, ctx) => {
				ctx.messages.length = 1;
				console.log("Context cleared.");
			},
		},
		{
			name: "exit",
			description: "Exit Heddle",
			execute: async (_args, ctx) => {
				console.log("Goodbye!");
				ctx.rl.close();
			},
		},
		{
			name: "quit",
			description: "Exit Heddle",
			execute: async (_args, ctx) => {
				console.log("Goodbye!");
				ctx.rl.close();
			},
		},
		{
			name: "cost",
			description: "Show token usage and cost",
			execute: async (_args, ctx) => {
				const { totalInputTokens, totalOutputTokens, totalCost } = ctx.costTracker;
				console.log(`  Input tokens:  ${totalInputTokens}`);
				console.log(`  Output tokens: ${totalOutputTokens}`);
				const costStr = totalCost !== null ? `$${totalCost.toFixed(4)}` : "N/A";
				console.log(`  Total cost:    ${costStr}`);
			},
		},
		{
			name: "status",
			description: "Show session status",
			execute: async (_args, ctx) => {
				console.log(`  Model:         ${ctx.config.model}`);
				console.log(`  Session:       ${ctx.sessionFile}`);
				console.log(`  Messages:      ${ctx.messages.length}`);
				console.log(`  Approval mode: ${ctx.config.approvalMode ?? "none"}`);
			},
		},
		{
			name: "context",
			description: "Show context size estimate",
			execute: async (_args, ctx) => {
				const totalChars = ctx.messages.reduce((sum, m) => {
					const content = typeof m.content === "string" ? m.content : "";
					return sum + content.length;
				}, 0);
				const estimatedTokens = Math.ceil(totalChars / 4);
				console.log(`  Messages:         ${ctx.messages.length}`);
				console.log(`  Estimated tokens: ~${estimatedTokens}`);
			},
		},
		{
			name: "model",
			description: "Switch model (e.g., /model openrouter/free)",
			execute: async (args, ctx) => {
				if (!args.trim()) {
					console.log(`  Current model: ${ctx.config.model}`);
					return;
				}
				ctx.setModel(args.trim());
				console.log(`  Model switched to: ${args.trim()}`);
			},
		},
		{
			name: "tools",
			description: "List available tools",
			execute: async (_args, ctx) => {
				for (const tool of ctx.registry.all()) {
					console.log(`  ${tool.name} — ${tool.description}`);
				}
			},
		},
		{
			name: "history",
			description: "Show recent message history",
			execute: async (args) => {
				const parts = args.trim().split(/\s+/).filter(Boolean);
				let limit = 20;
				let search: string | undefined;
				for (let i = 0; i < parts.length; i++) {
					const part = parts[i] as string;
					const next = parts[i + 1];
					if (part === "--limit" && next) {
						limit = Number.parseInt(next, 10) || 20;
						i++;
					} else if (part === "--search" && next) {
						search = parts.slice(i + 1).join(" ");
						break;
					} else if (part) {
						search = parts.slice(i).join(" ");
						break;
					}
				}
				const entries = await loadHistory({ limit, search });
				if (entries.length === 0) {
					console.log("  No history entries found.");
					return;
				}
				for (const entry of entries) {
					const time = entry.timestamp.replace("T", " ").replace(/\.\d+Z$/, "Z");
					console.log(`  [${time}] (${entry.content_type}) ${entry.message_preview}`);
				}
			},
		},
		{
			name: "restore",
			description: "Restore a file from backup (usage: /restore <file> [version])",
			execute: async (args, _ctx) => {
				const parts = args.trim().split(/\s+/);
				const filePath = parts[0];
				if (!filePath) {
					console.log("  Usage: /restore <file-path> [version]");
					return;
				}
				const { listBackups, restoreBackup } = await import("../file-history/restore.ts");
				if (parts[1]) {
					const result = await restoreBackup(filePath, Number(parts[1]));
					console.log(`  ${result}`);
				} else {
					const backups = await listBackups(filePath);
					if (backups.length === 0) {
						console.log(`  No backups found for ${filePath}`);
						return;
					}
					console.log(`  Backups for ${filePath}:`);
					for (const b of backups.slice(0, 10)) {
						console.log(`    v${b.version} — ${b.size} bytes`);
					}
					console.log("  Use /restore <file> <version> to restore");
				}
			},
		},
		{
			name: "compact",
			description: "Compact conversation context",
			execute: async (_args, ctx) => {
				if (!ctx.weakProvider) {
					console.log("  No weak model configured — cannot compact.");
					return;
				}
				const { compactContext } = await import("../context/compaction.ts");
				const modelLimit = ctx.config.maxTokens ?? 128000;
				const stats = await compactContext(ctx.messages, ctx.weakProvider, modelLimit);
				console.log(`  Compacted: removed ${stats.messagesRemoved} messages`);
				console.log(`  Tokens: ${stats.tokensBefore} → ${stats.tokensAfter}`);
			},
		},
		{
			name: "sessions",
			description: "List recent sessions",
			execute: async (_args, _ctx) => {
				const { listSessions } = await import("../session/list.ts");
				const sessions = await listSessions();
				if (sessions.length === 0) {
					console.log("  No sessions found.");
					return;
				}
				for (const s of sessions.slice(0, 20)) {
					const name = s.name ? ` (${s.name})` : "";
					const preview = s.firstUserMessage ? ` — ${s.firstUserMessage}` : "";
					console.log(`  ${s.id.slice(0, 8)}${name} | ${s.created} | ${s.messageCount} msgs${preview}`);
				}
			},
		},
		{
			name: "name",
			description: "Name the current session",
			execute: async (args, ctx) => {
				if (!args.trim()) {
					console.log("  Usage: /name <session-name>");
					return;
				}
				const { appendContextMarker } = await import("../session/jsonl.ts");
				await appendContextMarker(ctx.sessionFile, {
					type: "session_name",
					name: args.trim(),
					timestamp: new Date().toISOString(),
				});
				console.log(`  Session named: ${args.trim()}`);
			},
		},
		{
			name: "fork",
			description: "Fork the current session",
			execute: async (_args, ctx) => {
				const { forkSession } = await import("../session/fork.ts");
				const result = await forkSession(ctx.sessionFile);
				console.log(`  Forked to: ${result.sessionFile}`);
				console.log(`  New session ID: ${result.sessionId}`);
			},
		},
		{
			name: "tasks",
			description: "List or clear tracked tasks (usage: /tasks [clear])",
			execute: async (args, ctx) => {
				if (args.trim() === "clear") {
					const tasks = await loadTasks();
					const remaining = tasks.filter((t) => t.status !== "done");
					await saveTasks(remaining);
					console.log(`  Cleared ${tasks.length - remaining.length} completed tasks.`);
					return;
				}
				const tasks = await loadTasks();
				if (tasks.length === 0) {
					console.log("  No tasks tracked.");
					return;
				}
				console.log(formatTasksSummary(tasks, ctx.sessionId));
			},
		},
		{
			name: "agents",
			description: "List available agent definitions",
			execute: async (_args, ctx) => {
				if (ctx.agentDefinitions.size === 0) {
					console.log("  No agent definitions found.");
					return;
				}
				for (const [name, def] of ctx.agentDefinitions) {
					const model = def.model ? ` (${def.model})` : "";
					console.log(`  ${name}${model} — ${def.description}`);
				}
			},
		},
		{
			name: "plan",
			description: "Load or list saved plans (usage: /plan list | /plan load <name>)",
			execute: async (args) => {
				const parts = args.trim().split(/\s+/);
				const sub = parts[0];
				if (sub === "list" || !sub) {
					const plans = await listPlans();
					if (plans.length === 0) {
						console.log("  No saved plans.");
						return;
					}
					for (const p of plans) {
						const date = p.created ? p.created.replace("T", " ").slice(0, 19) : "unknown";
						console.log(`  ${p.name} (${date}) — ${p.preview}`);
					}
					return;
				}
				if (sub === "load") {
					const name = parts.slice(1).join(" ");
					if (!name) {
						console.log("  Usage: /plan load <name>");
						return;
					}
					const plan = await loadPlan(name);
					if (!plan) {
						console.log(`  Plan not found: "${name}"`);
						return;
					}
					console.log(`  Plan: ${plan.name}`);
					if (plan.meta.created) console.log(`  Created: ${plan.meta.created}`);
					if (plan.meta.model) console.log(`  Model: ${plan.meta.model}`);
					console.log(`\n${plan.content}`);
					return;
				}
				console.log("  Usage: /plan list | /plan load <name>");
			},
		},
		{
			name: "stats",
			description: "Show usage stats (usage: /stats [project])",
			execute: async (args, ctx) => {
				if (args.trim() === "project") {
					const { aggregateUsage } = await import("../usage/reader.ts");
					const { getProjectDir } = await import("../config/paths.ts");
					const stats = await aggregateUsage(getProjectDir());
					console.log(`  Sessions:      ${stats.totalSessions}`);
					console.log(`  Input tokens:  ${stats.totalTokens.input}`);
					console.log(`  Output tokens: ${stats.totalTokens.output}`);
					console.log(`  Total cost:    $${stats.totalCost.toFixed(4)}`);
					if (Object.keys(stats.toolCalls).length > 0) {
						console.log("  Tool calls:");
						for (const [tool, count] of Object.entries(stats.toolCalls).sort(([, a], [, b]) => b - a)) {
							console.log(`    ${tool}: ${count}`);
						}
					}
					return;
				}
				const { totalInputTokens, totalOutputTokens, totalCost } = ctx.costTracker;
				console.log(`  Input tokens:  ${totalInputTokens}`);
				console.log(`  Output tokens: ${totalOutputTokens}`);
				const costStr = totalCost !== null ? `$${totalCost.toFixed(4)}` : "N/A";
				console.log(`  Total cost:    ${costStr}`);
			},
		},
		{
			name: "paste",
			description: "Manage paste cache (usage: /paste [list|clear])",
			execute: async (args, ctx) => {
				if (!ctx.pasteCache) {
					console.log("  Paste cache is disabled.");
					return;
				}
				const sub = args.trim();
				if (sub === "clear") {
					ctx.pasteCache.clear();
					console.log("  Paste cache cleared.");
					return;
				}
				const entries = ctx.pasteCache.list();
				if (entries.length === 0) {
					console.log("  Paste cache is empty.");
					return;
				}
				for (const e of entries) {
					const id = e.pasteId ? ` [paste:${e.pasteId}]` : "";
					const size = Buffer.byteLength(e.content, "utf-8");
					console.log(`  ${e.path} (${e.lines} lines, ${size} bytes)${id}`);
				}
			},
		},
		{
			name: "agent",
			description: "Show agent definition details (usage: /agent <name>)",
			execute: async (args, ctx) => {
				const name = args.trim();
				if (!name) {
					console.log("  Usage: /agent <name>");
					console.log("  Use /agents to list available definitions.");
					return;
				}
				const def = ctx.agentDefinitions.get(name);
				if (!def) {
					console.log(`  Agent not found: "${name}"`);
					const available = [...ctx.agentDefinitions.keys()].join(", ");
					if (available) console.log(`  Available: ${available}`);
					return;
				}
				console.log(`  Name:        ${def.name}`);
				console.log(`  Description: ${def.description}`);
				if (def.model) console.log(`  Model:       ${def.model}`);
				if (def.tools) console.log(`  Tools:       ${def.tools.join(", ")}`);
				console.log(`  Source:      ${def.source}`);
				if (def.systemPrompt) {
					console.log(`  Prompt:      ${def.systemPrompt.slice(0, 200)}${def.systemPrompt.length > 200 ? "..." : ""}`);
				}
			},
		},
	];
}

File: src/commands/builtins.rs
//! Built-in slash commands shown by `/help`.

use std::sync::Arc;

use chrono::Utc;
use serde_json::json;

use super::types::{CommandContext, SlashCommand};
use crate::config::paths::get_project_dir;
use crate::context::compaction::{compact_context, CompactionConfig};
use crate::file_history::restore::{list_backups, restore_backup};
use crate::history::reader::{load_history, LoadHistoryOptions};
use crate::plans::storage::{list_plans, load_plan};
use crate::session::fork::{fork_session, ForkOptions};
use crate::session::jsonl::append_context_marker;
use crate::session::list::list_sessions;
use crate::tasks::storage::{format_tasks_summary, load_tasks, save_tasks};
use crate::usage::reader::aggregate_usage;

fn cmd<F>(name: &str, description: &str, exec: F) -> SlashCommand
where
    F: for<'a> Fn(
            &'a str,
            &'a mut CommandContext<'_>,
        ) -> std::pin::Pin<
            Box<dyn std::future::Future<Output = Option<String>> + Send + 'a>,
        > + Send
        + Sync
        + 'static,
{
    SlashCommand {
        name: name.to_string(),
        description: description.to_string(),
        execute: Arc::new(exec),
    }
}

pub fn create_builtin_commands() -> Vec<SlashCommand> {
    let mut out = Vec::new();

    out.push(cmd("help", "List available commands", |_args, _ctx| {
        Box::pin(async move {
            println!("  /help — see registered commands via /tools or /agents");
            None
        })
    }));

    out.push(cmd("clear", "Clear conversation context", |_args, ctx| {
        Box::pin(async move {
            ctx.messages.truncate(1);
            println!("Context cleared.");
            None
        })
    }));

    out.push(cmd("exit", "Exit Heddle", |_args, _ctx| {
        Box::pin(async move {
            println!("Goodbye!");
            std::process::exit(0);
        })
    }));
    out.push(cmd("quit", "Exit Heddle", |_args, _ctx| {
        Box::pin(async move {
            println!("Goodbye!");
            std::process::exit(0);
        })
    }));

    out.push(cmd("cost", "Show token usage and cost", |_args, ctx| {
        Box::pin(async move {
            let tracker = ctx.cost_tracker.lock();
            println!("  Input tokens:  {}", tracker.total_input_tokens());
            println!("  Output tokens: {}", tracker.total_output_tokens());
            let cost_str = tracker
                .total_cost()
                .map(|c| format!("${c:.4}"))
                .unwrap_or_else(|| "N/A".to_string());
            println!("  Total cost:    {cost_str}");
            None
        })
    }));

    out.push(cmd("status", "Show session status", |_args, ctx| {
        Box::pin(async move {
            println!("  Model:         {}", ctx.config.model);
            println!("  Session:       {}", ctx.session_file.display());
            println!("  Messages:      {}", ctx.messages.len());
            println!(
                "  Approval mode: {}",
                ctx.config
                    .approval_mode
                    .map(|m| m.as_str().to_string())
                    .unwrap_or_else(|| "none".to_string())
            );
            None
        })
    }));

    out.push(cmd(
        "context",
        "Show context size estimate",
        |_args, ctx| {
            Box::pin(async move {
                let total: usize = ctx
                    .messages
                    .iter()
                    .map(|m| m.content_str().map(|s| s.len()).unwrap_or(0))
                    .sum();
                let est = total.div_ceil(4);
                println!("  Messages:         {}", ctx.messages.len());
                println!("  Estimated tokens: ~{est}");
                None
            })
        },
    ));

    out.push(cmd(
        "model",
        "Switch model (e.g., /model openrouter/free)",
        |args, ctx| {
            Box::pin(async move {
                let trimmed = args.trim();
                if trimmed.is_empty() {
                    println!("  Current model: {}", ctx.config.model);
                    return None;
                }
                println!("  Model switched to: {trimmed}");
                Some(trimmed.to_string())
            })
        },
    ));

    out.push(cmd("tools", "List available tools", |_args, ctx| {
        Box::pin(async move {
            for tool in ctx.registry.all() {
                println!("  {} — {}", tool.name(), tool.description());
            }
            None
        })
    }));

    out.push(cmd(
        "history",
        "Show recent message history",
        |args, _ctx| {
            Box::pin(async move {
                let parts: Vec<&str> = args.split_whitespace().collect();
                let mut limit: usize = 20;
                let mut search: Option<String> = None;
                let mut i = 0;
                while i < parts.len() {
                    if parts[i] == "--limit" && i + 1 < parts.len() {
                        limit = parts[i + 1].parse().unwrap_or(20);
                        i += 2;
                    } else if parts[i] == "--search" && i + 1 < parts.len() {
                        search = Some(parts[i + 1..].join(" "));
                        break;
                    } else {
                        search = Some(parts[i..].join(" "));
                        break;
                    }
                }
                let entries = load_history(&LoadHistoryOptions {
                    limit: Some(limit),
                    search,
                });
                if entries.is_empty() {
                    println!("  No history entries found.");
                    return None;
                }
                for e in entries {
                    let time = e.timestamp.replace('T', " ");
                    println!("  [{time}] {:?} {}", e.content_type, e.message_preview);
                }
                None
            })
        },
    ));

    out.push(cmd(
        "restore",
        "Restore a file from backup (usage: /restore <file> [version])",
        |args, _ctx| {
            Box::pin(async move {
                let parts: Vec<&str> = args.split_whitespace().collect();
                let file_path = match parts.first() {
                    Some(p) => *p,
                    None => {
                        println!("  Usage: /restore <file-path> [version]");
                        return None;
                    }
                };
                if let Some(v) = parts.get(1) {
                    let version: u32 = v.parse().unwrap_or(0);
                    println!(
                        "  {}",
                        restore_backup(std::path::Path::new(file_path), version, None)
                    );
                } else {
                    let backups = list_backups(std::path::Path::new(file_path), None);
                    if backups.is_empty() {
                        println!("  No backups found for {file_path}");
                        return None;
                    }
                    println!("  Backups for {file_path}:");
                    for b in backups.iter().take(10) {
                        println!("    v{} — {} bytes", b.version, b.size);
                    }
                    println!("  Use /restore <file> <version> to restore");
                }
                None
            })
        },
    ));

    out.push(cmd(
        "compact",
        "Compact conversation context",
        |_args, ctx| {
            Box::pin(async move {
                let weak = match &ctx.weak_provider {
                    Some(w) => w.clone(),
                    None => {
                        println!("  No weak model configured — cannot compact.");
                        return None;
                    }
                };
                let model_limit = ctx.config.max_tokens.unwrap_or(128_000);
                match compact_context(
                    ctx.messages,
                    weak.as_ref(),
                    model_limit,
                    CompactionConfig::default(),
                )
                .await
                {
                    Ok(stats) => {
                        println!("  Compacted: removed {} messages", stats.messages_removed);
                        println!("  Tokens: {} → {}", stats.tokens_before, stats.tokens_after);
                    }
                    Err(e) => println!("  Error: {e}"),
                }
                None
            })
        },
    ));

    out.push(cmd("sessions", "List recent sessions", |_args, _ctx| {
        Box::pin(async move {
            let sessions = list_sessions(None);
            if sessions.is_empty() {
                println!("  No sessions found.");
                return None;
            }
            for s in sessions.iter().take(20) {
                let name = s
                    .name
                    .as_deref()
                    .map(|n| format!(" ({n})"))
                    .unwrap_or_default();
                let preview = s
                    .first_user_message
                    .as_deref()
                    .map(|m| format!(" — {m}"))
                    .unwrap_or_default();
                println!(
                    "  {}{name} | {} | {} msgs{preview}",
                    &s.id[..8.min(s.id.len())],
                    s.created,
                    s.message_count
                );
            }
            None
        })
    }));

    out.push(cmd("name", "Name the current session", |args, ctx| {
        Box::pin(async move {
            let name = args.trim();
            if name.is_empty() {
                println!("  Usage: /name <session-name>");
                return None;
            }
            let _ = append_context_marker(
                &ctx.session_file,
                &json!({
                    "type": "session_name",
                    "name": name,
                    "timestamp": Utc::now().to_rfc3339(),
                }),
            );
            println!("  Session named: {name}");
            None
        })
    }));

    out.push(cmd("fork", "Fork the current session", |_args, ctx| {
        Box::pin(async move {
            match fork_session(&ctx.session_file, ForkOptions::default()) {
                Ok(result) => {
                    println!("  Forked to: {}", result.session_file.display());
                    println!("  New session ID: {}", result.session_id);
                }
                Err(e) => println!("  Error: {e}"),
            }
            None
        })
    }));

    out.push(cmd(
        "tasks",
        "List or clear tracked tasks (usage: /tasks [clear])",
        |args, ctx| {
            Box::pin(async move {
                if args.trim() == "clear" {
                    let tasks = load_tasks(None);
                    let total = tasks.len();
                    let remaining: Vec<_> = tasks
                        .into_iter()
                        .filter(|t| t.status != crate::tasks::types::TaskStatus::Done)
                        .collect();
                    let cleared = total - remaining.len();
                    let _ = save_tasks(&remaining, None);
                    println!("  Cleared {cleared} completed tasks.");
                    return None;
                }
                let tasks = load_tasks(None);
                if tasks.is_empty() {
                    println!("  No tasks tracked.");
                    return None;
                }
                println!("{}", format_tasks_summary(&tasks, &ctx.session_id));
                None
            })
        },
    ));

    out.push(cmd(
        "agents",
        "List available agent definitions",
        |_args, ctx| {
            Box::pin(async move {
                if ctx.agent_definitions.is_empty() {
                    println!("  No agent definitions found.");
                    return None;
                }
                for (name, def) in ctx.agent_definitions {
                    let model = def
                        .model
                        .as_deref()
                        .map(|m| format!(" ({m})"))
                        .unwrap_or_default();
                    println!("  {name}{model} — {}", def.description);
                }
                None
            })
        },
    ));

    out.push(cmd(
        "plan",
        "Load or list saved plans (usage: /plan list | /plan load <name>)",
        |args, _ctx| {
            Box::pin(async move {
                let parts: Vec<&str> = args.split_whitespace().collect();
                let sub = parts.first().copied().unwrap_or("");
                if sub == "list" || sub.is_empty() {
                    let plans = list_plans(None);
                    if plans.is_empty() {
                        println!("  No saved plans.");
                        return None;
                    }
                    for p in plans {
                        let date = if p.created.is_empty() {
                            "unknown".to_string()
                        } else {
                            p.created.replace('T', " ").chars().take(19).collect()
                        };
                        println!("  {} ({date}) — {}", p.name, p.preview);
                    }
                    return None;
                }
                if sub == "load" {
                    let name = parts[1..].join(" ");
                    if name.is_empty() {
                        println!("  Usage: /plan load <name>");
                        return None;
                    }
                    match load_plan(&name, None) {
                        Some(p) => {
                            println!("  Plan: {}", p.name);
                            if let Some(c) = p.meta.get("created") {
                                println!("  Created: {c}");
                            }
                            if let Some(m) = p.meta.get("model") {
                                println!("  Model: {m}");
                            }
                            println!("\n{}", p.content);
                        }
                        None => println!("  Plan not found: {name:?}"),
                    }
                    return None;
                }
                println!("  Usage: /plan list | /plan load <name>");
                None
            })
        },
    ));

    out.push(cmd(
        "stats",
        "Show usage stats (usage: /stats [project])",
        |args, ctx| {
            Box::pin(async move {
                if args.trim() == "project" {
                    let stats = aggregate_usage(&get_project_dir(None));
                    println!("  Sessions:      {}", stats.total_sessions);
                    println!("  Input tokens:  {}", stats.total_input_tokens);
                    println!("  Output tokens: {}", stats.total_output_tokens);
                    println!("  Total cost:    ${:.4}", stats.total_cost);
                    if !stats.tool_calls.is_empty() {
                        println!("  Tool calls:");
                        let mut entries: Vec<_> = stats.tool_calls.iter().collect();
                        entries.sort_by(|a, b| b.1.cmp(a.1));
                        for (tool, count) in entries {
                            println!("    {tool}: {count}");
                        }
                    }
                    return None;
                }
                let tracker = ctx.cost_tracker.lock();
                println!("  Input tokens:  {}", tracker.total_input_tokens());
                println!("  Output tokens: {}", tracker.total_output_tokens());
                let cost_str = tracker
                    .total_cost()
                    .map(|c| format!("${c:.4}"))
                    .unwrap_or_else(|| "N/A".to_string());
                println!("  Total cost:    {cost_str}");
                None
            })
        },
    ));

    out.push(cmd(
        "paste",
        "Manage paste cache (usage: /paste [list|clear])",
        |args, ctx| {
            Box::pin(async move {
                let cache = match &ctx.paste_cache {
                    Some(c) => c.clone(),
                    None => {
                        println!("  Paste cache is disabled.");
                        return None;
                    }
                };
                let sub = args.trim();
                if sub == "clear" {
                    cache.lock().clear();
                    println!("  Paste cache cleared.");
                    return None;
                }
                let entries = cache.lock().list();
                if entries.is_empty() {
                    println!("  Paste cache is empty.");
                    return None;
                }
                for e in entries {
                    let id = e
                        .paste_id
                        .as_deref()
                        .map(|i| format!(" [paste:{i}]"))
                        .unwrap_or_default();
                    let size = e.content.len();
                    println!(
                        "  {} ({} lines, {size} bytes){id}",
                        e.path.display(),
                        e.lines
                    );
                }
                None
            })
        },
    ));

    out.push(cmd(
        "agent",
        "Show agent definition details (usage: /agent <name>)",
        |args, ctx| {
            Box::pin(async move {
                let name = args.trim();
                if name.is_empty() {
                    println!("  Usage: /agent <name>");
                    return None;
                }
                match ctx.agent_definitions.get(name) {
                    Some(def) => {
                        println!("  Name:        {}", def.name);
                        println!("  Description: {}", def.description);
                        if let Some(m) = &def.model {
                            println!("  Model:       {m}");
                        }
                        if let Some(t) = &def.tools {
                            println!("  Tools:       {}", t.join(", "));
                        }
                        println!("  Source:      {}", def.source.display());
                        if !def.system_prompt.is_empty() {
                            let preview: String = def.system_prompt.chars().take(200).collect();
                            let suffix = if def.system_prompt.len() > 200 {
                                "..."
                            } else {
                                ""
                            };
                            println!("  Prompt:      {preview}{suffix}");
                        }
                    }
                    None => {
                        println!("  Agent not found: {name:?}");
                    }
                }
                None
            })
        },
    ));

    out
}
